### PR TITLE
Address MISRA 12.1, 13, and 14

### PR DESCRIPTION
**Pull request by muneebahmed10**
*Issue #, if available:*

*Description of changes:*
Addresses violations of MISRA rules 12.1, 13.2, 13.3, 13.6, and 14.3 found by coverity.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/standard/mqtt/src/iot_mqtt_network.c [hunk 1]
```diff
@@ -483,13 +483,18 @@ static IotMqttError_t _deserializeIncomingPacket( _mqttConnection_t * pMqttConne
...
+
+    /* Packet validity is not checked when asserts are disabled. */
+    ( void ) packetValidity;
 
     /* A buffer for remaining data must be allocated if remaining length is not 0. */
     IotMqtt_Assert( ( pIncomingPacket->remainingLength > 0 ) ==
                     ( pIncomingPacket->pRemainingData != NULL ) );
 
     /* Only valid packets should be given to this function. */
-    IotMqtt_Assert( _incomingPacketValid( pIncomingPacket->type ) == true );
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/iot_mqtt_network.c]: The 13.6 violation here was "The operand of the sizeof operator contains potential side effect: calling function _incomingPacketValid(pIncomingPacket->type)". This seems to be a false positive as the `sizeof` operator shouldn't be getting called. I introduced a boolean here to reduce the number of coverity exceptions we have in comments, but that can be changed if it's more desirable.

File: libraries/standard/mqtt/src/iot_mqtt_serialize.c [hunk 2]
```diff
@@ -522,12 +522,14 @@ static uint8_t * _encodeUserName( uint8_t * pBuffer,
...
+                    *pBuffer = UINT16_HIGH_BYTE( ( pConnectInfo->userNameLength +
+                                                   AWS_IOT_METRICS_USERNAME_LENGTH ) );
+                    pBuffer++;
 
                     /* Write the low byte of the combined length. */
-                    *( pBuffer++ ) = UINT16_LOW_BYTE( ( pConnectInfo->userNameLength +
-                                                        AWS_IOT_METRICS_USERNAME_LENGTH ) );
+                    *pBuffer = UINT16_LOW_BYTE( ( pConnectInfo->userNameLength +
+                                                  AWS_IOT_METRICS_USERNAME_LENGTH ) );
+                    pBuffer++;
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/iot_mqtt_serialize.c]: This change does increase the complexity to 8, but this is acceptable as the requirement is <= 8 and not < 8.

File: libraries/standard/mqtt/src/iot_mqtt_network.c [hunk 3]
```diff
@@ -483,13 +483,18 @@ static IotMqttError_t _deserializeIncomingPacket( _mqttConnection_t * pMqttConne
                                                   _mqttPacket_t * pIncomingPacket )
 {
     IotMqttError_t status = IOT_MQTT_STATUS_PENDING;
+    bool packetValidity = false;
+
+    /* Packet validity is not checked when asserts are disabled. */
+    ( void ) packetValidity;
```
Comment (abhidixi11) [inline libraries/standard/mqtt/src/iot_mqtt_network.c]: packetValidity is used on line 496, which is not part of an assert statement, so (void) may not be required.

Comment (muneebahmed10) [reply libraries/standard/mqtt/src/iot_mqtt_network.c]: I was following the pattern I observed [here](https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta/libraries/standard/mqtt/src/iot_mqtt_operation.c#L717), but after testing with Travis, it doesn't look like there are any build warnings without the voids. I will remove them.

File: libraries/standard/common/include/iot_linear_containers.h [hunk 4]
```diff
@@ -496,6 +496,11 @@ static inline void IotListDouble_Remove( IotLink_t * const pLink )
     IotContainers_Assert( pLink != NULL );
 
     /* This function must be called on a linked element. */
+    /* Coverity finds a false positive in the below assert.
+     * As this is only an assert, the sizeof operator should
+     * not be called. Regardless, as IotLink_IsLinked does not
+     * mutate any data, there is no side effect in calling it. */
+    /* coverity[misra_c_2012_rule_13_6_violation] */
```
Comment (sarenameas) [inline libraries/standard/common/include/iot_linear_containers.h]: Does this violation still come up in the release type build?

Comment (muneebahmed10) [reply libraries/standard/common/include/iot_linear_containers.h]: No, it doesn't. It also doesn't come up if asserts are disabled.

Comment (gordonwang0) [reply libraries/standard/common/include/iot_linear_containers.h]: Let's remove the annotation then and run Coverity with asserts disabled.

File: libraries/standard/mqtt/src/iot_mqtt_network.c (hunk 1 above)
Comment (gordonwang0) [reply libraries/standard/mqtt/src/iot_mqtt_network.c]: Should change it back.  `packetValidity` is not used when asserts are disabled.

File: libraries/standard/mqtt/src/iot_mqtt_subscription.c [hunk 5]
```diff
@@ -526,7 +527,8 @@ void _IotMqtt_InvokeSubscriptionCallback( _mqttConnection_t * pMqttConnection,
         if( pSubscription->unsubscribed == true )
         {
             /* An unsubscribed subscription should have been removed from the list. */
-            IotMqtt_Assert( IotLink_IsLinked( &( pSubscription->link ) ) == false );
+            linkStatus = IotLink_IsLinked( &( pSubscription->link ) );
```
Comment (gordonwang0) [inline libraries/standard/mqtt/src/iot_mqtt_subscription.c]: Similarly, linkStatus will not be used when asserts are disabled.  Strange that the build check does not fail this...

File: libraries/standard/common/include/iot_linear_containers.h (hunk 4 above)
Comment (muneebahmed10) [reply libraries/standard/common/include/iot_linear_containers.h]: Reverted

File: libraries/standard/mqtt/src/iot_mqtt_network.c (hunk 1 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/src/iot_mqtt_network.c]: Reverted

**Pull request merged**